# Abstract

This paper applies the UniDip algorithm to the problem of genomic motif discovery. UniDip is a noise robust clustering algorithm based on the Hartigan's statistical test for unimodality, commonly called the dip-test. UniDip is different from other algorithms because it starts from the assumption that most data are noise only isolating regions of high density. Our work involves transforming symbolic genomic sequences into a data format compatible with UniDip's input requirements, including handling mis-alignments, mutations, and spurious motifs. We find that UniDip is beneficial both directly and indirectly to the problem of motif discovery. Directly UniDip is able to isolate multi-letter degenerate motifs in symbolic data provided the motif instances are aligned. Indirectly, and in conjunction with global alignment tools, UniDip is able to detect regions of higher conservation with greater concentrations of motif instances, which allow considerable time savings by focusing the search of alternative motif discovery algorithms.
